AR::RecordNotSaved & RecordNotDestroyed from save!/destroy! should include an error message

yuki24 opened the pull request:
When calling `AR::Base.save!` or `AR::Base.destroy!` and raising one of the exceptions, give a meaningful error message so that people can easily understand why it's failing to save/destroy.

cc @arthurnn @rafaelfranca

rafaelfranca commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -139,7 +139,8 @@ def save(*)
     def save!(*)
-      create_or_update || raise(RecordNotSaved.new(nil, self))
+      create_or_update || raise(RecordNotSaved.new("Failed to save the record " \
+                                                   "because one of the before callbacks returned false", self))
```
At master callbacks halt the execution using `throw(:abort)` we need to update the message now.

yuki24 replied:
ok, I'll change it to just `Failed to save the record`.

yuki24 committed:
AR::RecordNotSaved & RecordNotDestroyed should include an error message

When `AR::Base.save!` or `AR::Base.destroy!` is called and an exception
is raised, the exception doesn't have any error message or has a weird
message like `#<FailedBulb:0x0000000907b4b8>`. Give a better message so
we can easily understand why it's failing to save/destroy.

yuki24 commented:
updated the message and rebased. Now ready for merging.

arthurnn commented:
LGTM, @rafaelfranca final thoughts?

yuki24 commented:
@rafaelfranca any words on this?

yuki24 commented:
Is there any reason this pull request hasn't been merged yet?

yuki24 commented:
I ran into this issue today where the message from `ActiveRecord::RecordNotDestroyed` is competely broken.

```ruby
error = begin
  User.find(id).destroy!
rescue ActiveRecord::RecordNotDestroyed => e
  e
end

error.message
# => #<User:0x00000004e82f20>
```

This is still happening on Rails 4.2 and 5.0. I don't think nothing stops from merging this request and it should be merged and backported.

cc/ @rafaelfranca

rafaelfranca commented:
Backported at c61b17d

yuki24 commented:
@rafaelfranca Thanks! :heart: :green_heart: :yellow_heart:
